### PR TITLE
Feature/109672 create on start validation

### PR DESCRIPTION
**Pull request by laurens-ritense**

### CODE REVIEW

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+) : ApplicationRunner {
+
+    override fun run(args: ApplicationArguments?) {
+        validateAllViewModels()
+    }
+
+    fun validateAllViewModels() {
+        for (viewModelLoader in getAllViewModelLoaders()) {
+            if (!validateViewModel(viewModelLoader)) {
+                throw IllegalArgumentException("ViewModelLoader ${viewModelLoader.javaClass.simpleName} is not valid for form ${viewModelLoader.getFormName()}!")
```
Comment (sirolf2009) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt:24]: Can you add to the message which keys were a mismatch?

**Review by sirolf2009 — CHANGES_REQUESTED** [review]

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.ritense.formviewmodel.domain.validation
+
+import com.ritense.form.domain.FormIoFormDefinition
+
+
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt:5]: remove newline

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+class FormDefinitionPrefixedKeysExtractor {
+    fun getPrefixedKeys(form: FormIoFormDefinition): List<String> {
+        return form.inputFields.filter {
+            it["key"].asText().contains(PREFIX)
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt:9]: remove dub "key"

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.ritense.form.service.impl.FormIoFormDefinitionService
+import com.ritense.formviewmodel.domain.ViewModel
+import com.ritense.formviewmodel.domain.ViewModelLoader
+import org.springframework.boot.ApplicationArguments
+import org.springframework.boot.ApplicationRunner
+import org.springframework.context.ApplicationContext
+import org.springframework.stereotype.Component
+
+@Component
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt:11]: remove component

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/autoconfigure/FormViewModelAutoConfiguration.kt [hunk 5]
```diff
@@ -63,4 +66,13 @@ class FormViewModelAutoConfiguration {
         loaders
     )
 
+    @Bean
+    @Order(Ordered.LOWEST_PRECEDENCE)
+    fun onStartUpViewModelValidator(
+        formIoFormDefinitionService: FormIoFormDefinitionService,
+        context: ApplicationContext
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/autoconfigure/FormViewModelAutoConfiguration.kt:73]: use same approach as commandHandling like
 fun commandDispatcher(handlers: List<CommandHandler<*, *>>): CommandDispatcher {

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/autoconfigure/FormViewModelAutoConfiguration.kt [hunk 6]
```diff
@@ -63,4 +66,13 @@ class FormViewModelAutoConfiguration {
         loaders
     )
 
+    @Bean
+    @Order(Ordered.LOWEST_PRECEDENCE)
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/autoconfigure/FormViewModelAutoConfiguration.kt:70]: order can go

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/ClassPropertyExtractor.kt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.ritense.formviewmodel.domain.validation
+
+import com.ritense.formviewmodel.domain.ViewModel
+
+class ClassPropertyExtractor {
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/ClassPropertyExtractor.kt:5]: ViewModel.extractProperties() is my suggestion. Extension function in kotlin i mean give it a try

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package com.ritense.formviewmodel.domain.validation
+
+import com.ritense.form.domain.FormIoFormDefinition
+
+
+class FormDefinitionPrefixedKeysExtractor {
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/FormDefinitionPrefixedKeysExtractor.kt:6]: rename to ....ViewModelKeysExtractor

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+class OnStartUpViewModelValidator(
+    private val formIoFormDefinitionService: FormIoFormDefinitionService,
+    private val context: ApplicationContext,
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt:14]: refactor

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+) : ApplicationRunner {
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt:15]: I like it

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt (hunk 1 above)
Comment (Tom-Ritense) [reply form-view-model/src/main/kotlin/com/ritense/formviewmodel/domain/validation/OnStartUpViewModelValidator.kt:24]: This should be really clear what is wrong. Also combine ALL errors in one list.

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/web/rest/FormViewModelResource.kt [hunk 11]
```diff
@@ -34,34 +34,34 @@ class FormViewModelResource(
...
     @PostMapping
     fun updateFormViewModel(
-        @RequestParam(required = true) formId: String,
+        @RequestParam(required = true) formName: String,
         @RequestParam(required = true) taskInstanceId: String,
         @RequestBody formViewModel: String
     ): ResponseEntity<ViewModel> {
         return ResponseEntity.ok(
-            parseViewModel(formViewModel, viewModelLoaderFactory.getViewModelLoader(formId)?.getViewModelType()!!).update()
+            parseViewModel(formViewModel, viewModelLoaderFactory.getViewModelLoader(formName)?.getViewModelType()!!).update()
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/web/rest/FormViewModelResource.kt:52]: !! seems a code smell type optional?

File: form-view-model/src/main/kotlin/com/ritense/formviewmodel/web/rest/FormViewModelResource.kt [hunk 12]
```diff
@@ -34,34 +34,34 @@ class FormViewModelResource(
 
     @GetMapping
     fun getFormViewModel(
-        @RequestParam(required = true) formId: String,
+        @RequestParam(required = true) formName: String,
         @RequestParam(required = true) taskInstanceId: String
     ): ResponseEntity<ViewModel?> {
         return ResponseEntity.ok(
-            viewModelLoaderFactory.getViewModelLoader(formId)?.onLoad(taskInstanceId)
+            viewModelLoaderFactory.getViewModelLoader(formName)?.onLoad(taskInstanceId)
```
Comment (Tom-Ritense) [inline form-view-model/src/main/kotlin/com/ritense/formviewmodel/web/rest/FormViewModelResource.kt:41]: rename to load

**Review by Tom-Ritense — CHANGES_REQUESTED** [review]

**Comment by Tom-Ritense** [comment]
old

**Pull request closed without merging**